### PR TITLE
ARROW-9160: [C++] Implement contains for exact matches

### PR DESCRIPTION
**Pull request by xhochy**

### CODE REVIEW

**Comment by xhochy** [comment]
This currently fails for chunked arrays. I though that they should be handled by the kernel framework automatically but it seems, they aren't.

```
[----------] 1 test from TestStringKernels/1, where TypeParam = arrow::LargeStringType
[ RUN      ] TestStringKernels/1.ContainsExact
../src/arrow/testing/gtest_util.cc:149: Failure
Failed
Got:
  [
    [
      false,
      null,
      false
    ]
  ]
Expected:
  [
    [
      true,
      null
    ],
    [
      false
    ]
  ]
../src/arrow/testing/gtest_util.cc:149: Failure
Failed
Got:
  [
    [
      false,
      false,
      false,
      null,
      false
    ]
  ]
Expected:
  [
    [
      true,
      false
    ],
    [
      true,
      null,
      false
    ]
  ]
[  FAILED  ] TestStringKernels/1.ContainsExact, where TypeParam = arrow::LargeStringType (2 ms)
```

@wesm @pitrou Any pointers what I'm missing?

**Comment by wesm** [comment]
@xhochy chunked array should be handled automatically by the function executors. I will take a look.

**Comment by xhochy** [comment]
Implemented Knuth-Morris-Pratt, got roughly a 2x speedup in the benchmark.

```
Before:
ContainsExact   53673203 ns     53323692 ns           13 bytes_per_second=297.061M/s items_per_second=19.6644M/s
After:
ContainsExact   30150889 ns     29982087 ns           23 bytes_per_second=528.329M/s items_per_second=34.9734M/s
```

**Comment by wesm** [comment]
@xhochy I'm fixing a couple issues with the implementation:

* The function executor allocates memory for you unless you explicitly disable it. The idea is that you don't want to do allocations in-kernel unless it is necessary. The preallocation settings affect how execution with chunked arrays work
* The slice offset of the `Datum* out` parameter must be respected (it may be non-zero -- the executor does not preserve the chunk layout of ChunkedArrays and will merge chunks to make larger output arrays -- this can be configured in `ExecContext`)

File: cpp/src/arrow/compute/kernels/scalar_string.cc [hunk 1]
```diff
@@ -387,16 +380,20 @@ using ContainsExactState = OptionsWrapper<ContainsExactOptions>;
...
+    const uint8_t* pat = reinterpret_cast<const uint8_t*>(arg.pattern.c_str());
+    const int64_t pat_size = arg.pattern.length();
+    StringBoolTransform<Type>(
+        ctx, batch,
+        [pat, pat_size](const void* offsets, const uint8_t* data, int64_t length,
+                        int64_t output_offset, uint8_t* output) {
+          TransformContainsExact<offset_type>(
+              pat, pat_size, reinterpret_cast<const offset_type*>(offsets), data, length,
+              output_offset, output);
+        },
```
Comment (wesm) [inline cpp/src/arrow/compute/kernels/scalar_string.cc:395]: I hope you don't mind, this seemed simpler to me

**Comment by wesm** [comment]
I just opened https://issues.apache.org/jira/browse/ARROW-9285 -- it should be easy to check if a kernel has mistakenly replaced a preallocated data buffer (which may be a slice of a larger output that is being populated)

**Comment by wesm** [comment]
I also propose to rename the function from "contains_exact" to "utf8_contains". I'm pushing that change here shortly

**Comment by xhochy** [comment]
I think prefixes make sense. We will have sometime similar kernel names that act quite different depending on the types they work on. I would differentiate in the string/binary case the kernels between the categories:

* `binary_`: works on any binary data
* `ascii_`: works on string data but is limited ASCII-only characters
* `utf8_`: works on string data and respects UTF8-encoding or even makes use of an UTF8-symbol table (i.e. `utf8proc`).

With that, I would rename the kernel here to `binary_contains_exact`. The `exact` suffix is important for me as I intent to implement also `ut8_contains_case_insensitve`, `binary_contains_regex` (or should this be string/utf8?), `utf8_contains_regex_case_insensitve`.

**Comment by pitrou** [comment]
You match a regex, you don't contain it.

**Comment by xhochy** [comment]
> You match a regex, you don't contain it.

That is one of the name clashes, we already have a match kernel.

**Comment by pitrou** [comment]
`match_regex` then? :-)

**Comment by maartenbreddels** [comment]
I like the prefixing by `string`. I'm a big fan of ordering 'words' in snake or camel casing for good tab completion and alphabetic ordering, so I agree with @wesm 's proposal. As a user, I first think of the type I work with ('string'), next, what I want to do with it (upper/lower casing). Tab-completion/ordering would then reveal 2 variants, very intuitive.
To make this concrete:
 * string_lower_utf8
 * string_lower_ascii
 * binary_contains_exact
 * string_contains_regex_ascii
 * string_contains_regex_utf8

(or s/contains/match 😄 )

When working with strings, you need to check 

There might be kernels that work on binary data, but do not work well with utf8, e.g. a `binary_slice`, which would slice each binary-string on a byte basis. It could cut a multibyte encoded codepoint to result in an invalid utf8 string. I'd say that's acceptable, we could check that depending on the type we get it.

Regarding case insensitive variants, I think we should expose functions to do normalization (eg replace the single codepoint ë by the letter e and the diaeresis combining character(the dots above the ë)), case folding, and removal of combining characters. That allows users to remove e.g. the diaeresis from ë to do pattern matching without diacritics.

**Comment by maartenbreddels** [comment]
While at the difficult topic of naming, is there a conversion (agreed or emerging) for naming the functors/ArrayKernelExec implementations? I see in `scalar_string.cc` we use Transform, which is probably redundant (as I understand scalar kernels basically map/transform elementwise). But I think it would be useful to include the type name, e.g.
 * Utf8ToUtf8 (which is now Utf8Transform) (hence `Utf8Upper : Utf8ToUtf8<Type, Utf8Upper<Type>>`)
 * Utf8ToBool / Utf8Predicate (which is now StringBoolTransform)
 * Utf8ToScalar
 * BinaryToBool

**Comment by maartenbreddels** [comment]
> * The function executor allocates memory for you unless you explicitly disable it. The idea is that you don't want to do allocations in-kernel unless it is necessary. The preallocation settings affect how execution with chunked arrays work

Could you elaborate? Why is this not a problem with the lower/upper kernels?

File: cpp/src/arrow/compute/kernels/scalar_string.cc [hunk 2]
```diff
@@ -297,6 +297,116 @@ void AddAsciiLength(FunctionRegistry* registry) {
...
+          bitmap_writer.Set();
+          break;
+        }
+      } else {
+        pattern_pos = std::max<offset_type>(0, prefix_table[pattern_pos]);
+      }
+    }
+    bitmap_writer.Next();
+  }
+  bitmap_writer.Finish();
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/scalar_string.cc:376]: It is me, or you're never clearing the output bit when the pattern isn't found?

File: cpp/src/arrow/compute/api_scalar.h [hunk 3]
```diff
@@ -259,6 +259,15 @@ Result<Datum> IsValid(const Datum& values, ExecContext* ctx = NULLPTR);
 ARROW_EXPORT
 Result<Datum> IsNull(const Datum& values, ExecContext* ctx = NULLPTR);
 
+// ----------------------------------------------------------------------
+// String functions
+
+struct ARROW_EXPORT ContainsExactOptions : public FunctionOptions {
+  explicit ContainsExactOptions(std::string pattern = "") : pattern(pattern) {}
```
Comment (pitrou) [inline cpp/src/arrow/compute/api_scalar.h:266]: I don't think defaulting to the empty string makes sense here.

File: python/pyarrow/compute.py [hunk 4]
```diff
@@ -113,6 +113,24 @@ def func(left, right):
 multiply = _simple_binary_function('multiply')
 
 
+def contains_exact(array, pattern):
+    """
+    Check whether a pattern occurs as part of the values of the array.
```
Comment (pitrou) [inline python/pyarrow/compute.py:118]: That's rather vague. Does `contains_exact(pa.array([1,2,3]), pa.array([1,2]))` return true? I don't think so.

**Review by pitrou — COMMENTED** [review]
This is a nice addition. A couple comments.

**Comment by wesm** [comment]
> Could you elaborate? Why is this not a problem with the lower/upper kernels?

The data preallocation is only for fixed size outputs (eg boolean, integers, floating point, etc).

File: cpp/src/arrow/compute/kernels/scalar_string.cc [hunk 5]
```diff
@@ -297,6 +297,116 @@ void AddAsciiLength(FunctionRegistry* registry) {
...
+    ArrayData* out_arr = out->mutable_array();
+    if (input.length > 0) {
+      transform(
+          reinterpret_cast<const offset_type*>(input.buffers[1]->data()) + input.offset,
+          input.buffers[2]->data(), input.length, out_arr->offset,
+          out_arr->buffers[1]->mutable_data());
+    }
+  } else {
+    const auto& input = checked_cast<const BaseBinaryScalar&>(*batch[0].scalar());
+    auto result = checked_pointer_cast<BooleanScalar>(MakeNullScalar(out->type()));
```
Comment (kiszk) [inline cpp/src/arrow/compute/kernels/scalar_string.cc:324]: We can move this line into then clause before line 327.

File: cpp/src/arrow/compute/api_scalar.h (hunk 3 above)
Comment (xhochy) [reply cpp/src/arrow/compute/api_scalar.h:266]: This is needed to be able to construct an instance of it in Cython. I can remove this and switch the Cython instantiation to a pointer.

File: cpp/src/arrow/compute/kernels/scalar_string.cc (hunk 2 above)
Comment (xhochy) [reply cpp/src/arrow/compute/kernels/scalar_string.cc:376]: The bit vector is 0-initialised, so this is not needed (`FirstTimeBitmapWriter::Clear` is no-op).

Comment (pitrou) [reply cpp/src/arrow/compute/kernels/scalar_string.cc:376]: I forgot about that, thanks.

**Comment by wesm** [comment]
> I think prefixes make sense. We will have sometime similar kernel names that act quite different depending on the types they work on. I would differentiate in the string/binary case the kernels between the categories:
> 
> * `binary_`: works on any binary data
> * `ascii_`: works on string data but is limited ASCII-only characters
> * `utf8_`: works on string data and respects UTF8-encoding or even makes use of an UTF8-symbol table (i.e. `utf8proc`).
> 
> With that, I would rename the kernel here to `binary_contains_exact`. The `exact` suffix is important for me as I intent to implement also `ut8_contains_case_insensitve`, `binary_contains_regex` (or should this be string/utf8?), `utf8_contains_regex_case_insensitve`.

This seems reasonable to me

**Comment by pitrou** [comment]
"binary_contains_exact" still works for utf8 data, though.

**Comment by wesm** [comment]
OK, do you have an idea for a prefix for a function that works either on BinaryArray or StringArray?

**Comment by maartenbreddels** [comment]
String is a subclass of binary right? So I expect binary_contains_exact to work on strings right?

**Comment by pitrou** [comment]
Right, I was just mentioned that "binary" isn't exclusive of "utf8" in this particular instance.
However, I'm not sure why we need prefixes. "contains" seems sufficient to me :-)

**Comment by wesm** [comment]
Other types (e.g. list) will need to have a "contains" operation, so I think it's more clear to have `binary_contains` and `list_contains` than simply `contains`.

**Comment by xhochy** [comment]
Adressed all comments and CI is happy, too.

File: cpp/src/arrow/compute/kernels/scalar_string_benchmark.cc [hunk 6]
```diff
@@ -56,6 +57,11 @@ static void AsciiUpper(benchmark::State& state) {
   UnaryStringBenchmark(state, "ascii_upper");
 }
 
+static void BinaryContainsExact(benchmark::State& state) {
+  BinaryContainsExactOptions options("abac");
+  UnaryStringBenchmark(state, "contains_exact", &options);
```
Comment (wesm) [inline cpp/src/arrow/compute/kernels/scalar_string_benchmark.cc:62]: Looks like one rename missed here

**Review by wesm — APPROVED** [review]
+1. I'll fix the benchmark function rename and then merge this

**Pull request closed without merging**